hashring/hashring.go

```diff
@@ -123,43 +119,42 @@ func (r *HashRing) AddServer(address string) {
 	r.computeChecksum()
 }
 
-// inserts server replicas into ring
 func (r *HashRing) addReplicas(server string) {
 	r.servers.Lock()
-	r.servers.byAddress[server] = true
+	defer r.servers.Unlock()
 
+	r.servers.byAddress[server] = true
 	for i := 0; i < r.replicaPoints; i++ {
 		address := fmt.Sprintf("%s%v", server, i)
-		r.servers.tree.Insert(int(r.hashfunc([]byte(address))), server)
+		r.servers.tree.Insert(r.hashfunc(address), server)
 	}
-
-	r.servers.Unlock()
 }
 
+// RemoveServer removes all replicas of a server from the HashRing
 func (r *HashRing) RemoveServer(address string) {
 	if !r.HasServer(address) {
 		return
 	}
 
-	r.RemoveReplicas(address)
+	r.removeReplicas(address)
```

Comment:
What do you think about atomically removing the replicas and updating the checksums?  It would be a nice invariant to have that as long as you hold a lock, the list of servers and checksums match.